### PR TITLE
fix(NVIM_APPNAME): show error message if $NVIM_APPNAME is invalid

### LINKED ISSUE #23056: tempname() fails when NVIM_APPNAME is set to subdirectory
### Problem

When `NVIM_APPNAME` contains a subdirectory, tempname() fails to return anything. For example, I store all of my nvim configs in `$HOME/nvim-profiles` and use an alias to run nvim with the the `NVIM_APPNAME` variable set. However, any plugins that rely on `tempname()` to generate a temporary directory are not able to run properly. 

Before `NVIM_APPNAME`, I just set the various `XDG_HOME` variables using the following shell script, and everything worked fine:
<details>
<summary>Shell Script</summary>

```shell
#!/bin/sh
NVIM_PROFILE_NAME=""
NVIM_COMMAND="nvim"

exec env XDG_CONFIG_HOME="$HOME/.config/nvim-profiles/${NVIM_PROFILE_NAME}" \
         XDG_DATA_HOME="$HOME/.local/share/nvim-profiles/${NVIM_PROFILE_NAME}" \
         XDG_STATE_HOME="$HOME/.local/state/nvim-profiles/${NVIM_PROFILE_NAME}" \
         XDG_CACHE_HOME="$HOME/.cache/nvim-profiles/${NVIM_PROFILE_NAME}" \
         ${NVIM_COMMAND} \
         "$@"
```

</details>

### Steps to reproduce
1. Run the following commands in terminal
```shell
cd $HOME
NVIM_APPNAME="some_directory/TestVim" nvim
```
2. execute "echo tempname()" in neovim
3. Nothing gets returned
4. Run `NVIM_APPNAME="TestVim" nvim`
5. `tempname()` returns a temporary directory path

### Expected behavior

I would expect `tempname()` to return the path to a temporary directory:

Screen Recording 2023-04-12 at 1 50 51 PM

Interestingly enough, calling the alias in the `~/.local/state` directory allows tempname() to return a directory. No clue why this happens.

Screen_Recording_2023-04-12_at_12 39 51_PM

### Neovim version (nvim -v)

NVIM v0.10.0-dev-46+ge23c6ebed

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 13.3

### Terminal name/version

Kitty 0.27.1

### $TERM environment variable

xterm-kitty

### Installation

bob

### PR DESCRIPTION
**Pull request by glacambre**
Closes #23056 .

### CODE REVIEW

File: src/nvim/os/stdpaths.c [hunk 1]
```diff
@@ -69,6 +69,19 @@ const char *get_appname(void)
   return env_val;
 }
 
+/// Ensure that APPNAME is valid. In particular, it cannot contain directory separators.
```
Comment (zeertzjq) [inline src/nvim/os/stdpaths.c:72]:
```suggestion
/// Ensure that NVIM_APPNAME is valid. In particular, it cannot contain directory separators.
```

**Comment by glacambre** [comment]
The windows & OSX failures are not caused by the doc change nor by the presence of `$` in the error string. It has to come from the code. Perhaps the problem is that we're calling getenv() too early - before some initialization code has had the chance to run?

File: runtime/doc/starting.txt [hunk 2]
```diff
@@ -1392,6 +1392,8 @@ Neovim will look for (and auto-create) in the various XDG parent directories.
 For example, setting `$NVIM_APPNAME` to "neovim" before running Neovim will
 result in Neovim looking for configuration files in `$XDG_CONFIG_HOME/neovim`
 instead of `$XDG_CONFIG_HOME/nvim`.
+`$NVIM_APPNAME` should not contain any path separators or characters that are
```
Comment (justinmk) [inline runtime/doc/starting.txt:1395]:
```suggestion
`$NVIM_APPNAME` must not contain any path separators or characters that are
```

Alternatively I'd even just leave this out of the docs. It's guarded by an explicit check and it's obvious.

File: src/nvim/main.c [hunk 3]
```diff
@@ -250,6 +250,11 @@ int main(int argc, char **argv)
 
   argv0 = argv[0];
 
+  if (!appname_is_valid()) {
+    os_errmsg("Invalid $NVIM_APPNAME (see `:help $NVIM_APPNAME`)\n");
```
Comment (justinmk) [inline src/nvim/main.c:254]: "foo (see :help foo)" is redundant

we probably want a EXXX error code for this if we're going to mention it in the docs.

```suggestion
    os_errmsg("$NVIM_APPNAME must not contain path separators\n");
```

File: src/nvim/main.c [hunk 4]
```diff
@@ -250,6 +250,11 @@ int main(int argc, char **argv)
 
   argv0 = argv[0];
 
+  if (!appname_is_valid()) {
```
Comment (justinmk) [inline src/nvim/main.c:253]: no need to do this check so early. Instead do it _just_ before $NVIM_APPNAME is evaluated. That should help avoid order-of-operations issues.

**Comment by dundargoc** [comment]
@glacambre I think this PR should pass the tests if you rebase to master.

File: runtime/doc/starting.txt (hunk 2 above)
Comment (glacambre) [reply runtime/doc/starting.txt]: Very good point! Done.

File: src/nvim/main.c (hunk 3 above)
Comment (glacambre) [reply src/nvim/main.c]: I removed references to the documentation as suggested and attempted to make the requirements both broader and clearer (to cover cases where users might attempt to use characters forbidden by their filesystem, like `:` under NTFS).

File: src/nvim/main.c (hunk 4 above)
Comment (glacambre) [reply src/nvim/main.c:252]: The first call to get_appname is performed in main (src/nvim/main.c:286) -> early_init (src/nvim/main.c:222) -> set_init_1 (src/nvim/option.c:340) -> stdpaths_user_state_subpath  (src/nvim/os/stdpaths.c:202) -> get_xdg_home (src/nvim/os/stdpaths.c:138.). This doesn't seem to be much later than the current call to appname_is_valid and I'm not sure getting closer to this call would change anything (but I could very well be missing something).

**Pull request merged**